File: lib/services/templates.js
import db from '../database.js';

/**
 * Templates Service
 * Handles all template-related database operations
 */

export class TemplatesService {
  constructor() {
    this.filename = 'templates.json';
  }

  /**
   * Get all templates
   */
  async getAll() {
    try {
      const templates = await db.readData(this.filename, []);
      return templates.sort((a, b) => 
        new Date(b.updatedAt || b.createdAt).getTime() - new Date(a.updatedAt || a.createdAt).getTime()
      );
    } catch (error) {
      console.error('Error getting all templates:', error);
      return [];
    }
  }

  /**
   * Get active templates only
   */
  async getActive() {
    try {
      const templates = await this.getAll();
      return templates.filter(template => template.active !== false);
    } catch (error) {
      console.error('Error getting active templates:', error);
      return [];
    }
  }

  /**
   * Get template by ID
   */
  async getById(id) {
    try {
      const templates = await this.getAll();
      return templates.find(template => template.id === id) || null;
    } catch (error) {
      console.error('Error getting template by ID:', error);
      return null;
    }
  }

  /**
   * Create new template
   */
  async create(templateData) {
    try {
      const templates = await this.getAll();
      
      const newTemplate = {
        id: db.generateId('template'),
        createdAt: db.getTimestamp(),
        updatedAt: db.getTimestamp(),
        active: true,
        ...templateData
      };

      templates.push(newTemplate);
      await db.writeData(this.filename, templates);
      
      console.log('✅ Template created successfully:', newTemplate.id);
      return newTemplate;
    } catch (error) {
      console.error('Error creating template:', error);
      throw new Error('Failed to create template');
    }
  }

  /**
   * Update template
   */
  async update(id, updates) {
    try {
      const templates = await this.getAll();
      const index = templates.findIndex(template => template.id === id);
      
      if (index === -1) {
        throw new Error('Template not found');
      }

      templates[index] = { 
        ...templates[index], 
        ...updates,
        updatedAt: db.getTimestamp()
      };

      await db.writeData(this.filename, templates);
      
      console.log('✅ Template updated successfully:', id);
      return templates[index];
    } catch (error) {
      console.error('Error updating template:', error);
      throw new Error('Failed to update template');
    }
  }

  /**
   * Delete template
   */
  async delete(id) {
    try {
      const templates = await this.getAll();
      const filteredTemplates = templates.filter(template => template.id !== id);
      
      if (filteredTemplates.length === templates.length) {
        throw new Error('Template not found');
      }

      await db.writeData(this.filename, filteredTemplates);
      
      console.log('✅ Template deleted successfully:', id);
      return true;
    } catch (error) {
      console.error('Error deleting template:', error);
      throw new Error('Failed to delete template');
    }
  }

  /**
   * Search templates
   */
  async search(searchTerm) {
    try {
      const templates = await this.getAll();
      
      if (!searchTerm) return templates;

      const term = searchTerm.toLowerCase();
      return templates.filter(template =>
        template.name?.toLowerCase().includes(term) ||
        template.shortDesc?.toLowerCase().includes(term) ||
        template.longDesc?.toLowerCase().includes(term) ||
        template.features?.some(feature => feature.toLowerCase().includes(term)) ||
        template.useCases?.some(useCase => useCase.toLowerCase().includes(term))
      );
    } catch (error) {
      console.error('Error searching templates:', error);
      return [];
    }
  }
}

// Create and export a single instance
const templatesService = new TemplatesService();
export default templatesService;

File: app/services/templates.py
from typing import List, Dict, Any, Optional
from .database import db

class TemplatesService:
    def __init__(self):
        self.filename = 'templates.json'
    
    async def get_all(self) -> List[Dict[str, Any]]:
        """Get all templates"""
        try:
            templates = await db.read_data(self.filename, [])
            return templates
        except Exception as e:
            print(f"Error getting all templates: {e}")
            return []
    
    async def get_by_id(self, template_id: str) -> Optional[Dict[str, Any]]:
        """Get template by ID"""
        try:
            templates = await self.get_all()
            return next((tpl for tpl in templates if tpl.get('id') == template_id), None)
        except Exception as e:
            print(f"Error getting template by ID: {e}")
            return None
    
    async def create(self, template_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create new template"""
        try:
            templates = await self.get_all()
            
            new_template = {
                'id': db.generate_id('template'),
                'createdAt': db.get_timestamp(),
                **template_data
            }
            
            templates.append(new_template)
            await db.write_data(self.filename, templates)
            
            print(f"✅ Template created successfully: {new_template['id']}")
            return new_template
        except Exception as e:
            print(f"Error creating template: {e}")
            raise Exception("Failed to create template")
    
    async def update(self, template_id: str, updates: Dict[str, Any]) -> Dict[str, Any]:
        """Update template"""
        try:
            templates = await self.get_all()
            index = next((i for i, tpl in enumerate(templates) if tpl.get('id') == template_id), -1)
            
            if index == -1:
                raise Exception("Template not found")
            
            templates[index] = {
                **templates[index],
                **updates,
                'updatedAt': db.get_timestamp()
            }
            
            await db.write_data(self.filename, templates)
            print(f"✅ Template updated successfully: {template_id}")
            return templates[index]
        except Exception as e:
            print(f"Error updating template: {e}")
            raise Exception("Failed to update template")
    
    async def delete(self, template_id: str) -> bool:
        """Delete template"""
        try:
            templates = await self.get_all()
            filtered_templates = [tpl for tpl in templates if tpl.get('id') != template_id]
            
            if len(filtered_templates) == len(templates):
                raise Exception("Template not found")
            
            await db.write_data(self.filename, filtered_templates)
            print(f"✅ Template deleted successfully: {template_id}")
            return True
        except Exception as e:
            print(f"Error deleting template: {e}")
            raise Exception("Failed to delete template")

# Global service instance
templates_service = TemplatesService()
